File: engine/graph.cpp
#include "graph.h"
#include <fstream>
#include <sstream>
#include <algorithm>
#include <functional>

// ──────────────────────────────────────────────
// Internal helpers
// ──────────────────────────────────────────────
void Graph::ensureSize(int id) {
    if (id > maxId_) {
        maxId_ = id;
        adj_.resize(maxId_ + 1);
    }
}

// Trim whitespace
static std::string trim(const std::string& s) {
    size_t start = s.find_first_not_of(" \t\r\n");
    if (start == std::string::npos) return "";
    size_t end = s.find_last_not_of(" \t\r\n");
    return s.substr(start, end - start + 1);
}

// ──────────────────────────────────────────────
// Graph construction
// ──────────────────────────────────────────────
void Graph::addWarehouse(const Warehouse& w) {
    ensureSize(w.id);
    warehouses_[w.id] = w;
}

void Graph::addEdge(int from, int to, double cost, double time_hours) {
    ensureSize(from);
    ensureSize(to);
    // Bidirectional edges
    adj_[from].push_back({to, cost, time_hours});
    adj_[to].push_back({from, cost, time_hours});
}

// ──────────────────────────────────────────────
// Dijkstra's Algorithm — O((V+E) log V)
//   Uses min-heap (priority queue) to greedily
//   expand the cheapest frontier node.
// ──────────────────────────────────────────────
PathResult Graph::dijkstra(int src, int dst) const {
    PathResult result;
    result.reachable = false;
    result.total_cost = 0;
    result.total_time = 0;

    int n = (int)adj_.size();
    if (src < 0 || src >= n || dst < 0 || dst >= n) return result;

    const double INF = std::numeric_limits<double>::infinity();
    std::vector<double> dist(n, INF);
    std::vector<double> timeArr(n, INF);
    std::vector<int> prev(n, -1);

    // Min-heap: (cost, node)
    using PII = std::pair<double, int>;
    std::priority_queue<PII, std::vector<PII>, std::greater<PII>> pq;

    dist[src] = 0;
    timeArr[src] = 0;
    pq.push({0, src});

    while (!pq.empty()) {
        auto topPair = pq.top();
        double d = topPair.first;
        int u = topPair.second;
        pq.pop();

        if (d > dist[u]) continue;  // stale entry
        if (u == dst) break;        // found shortest

        for (const auto& edge : adj_[u]) {
            double newDist = dist[u] + edge.cost;
            if (newDist < dist[edge.to]) {
                dist[edge.to] = newDist;
                timeArr[edge.to] = timeArr[u] + edge.time_hours;
                prev[edge.to] = u;
                pq.push({newDist, edge.to});
            }
        }
    }

    if (dist[dst] == INF) return result;

    // Reconstruct path by backtracking through prev[]
    result.reachable = true;
    result.total_cost = dist[dst];
    result.total_time = timeArr[dst];

    int cur = dst;
    while (cur != -1) {
        result.path.push_back(cur);
        cur = prev[cur];
    }
    std::reverse(result.path.begin(), result.path.end());

    return result;
}

// ──────────────────────────────────────────────
// BFS — O(V+E) shortest path by hop count
//   Treats all edges as weight=1
// ──────────────────────────────────────────────
PathResult Graph::bfs(int src, int dst) const {
    PathResult result;
    result.reachable = false;
    result.total_cost = 0;
    result.total_time = 0;

    int n = (int)adj_.size();
    if (src < 0 || src >= n || dst < 0 || dst >= n) return result;

    std::vector<int> prev(n, -1);
    std::vector<bool> visited(n, false);
    std::queue<int> q;

    visited[src] = true;
    q.push(src);

    while (!q.empty()) {
        int u = q.front();
        q.pop();
        if (u == dst) break;

        for (const auto& edge : adj_[u]) {
            if (!visited[edge.to]) {
                visited[edge.to] = true;
                prev[edge.to] = u;
                q.push(edge.to);
            }
        }
    }

    if (!visited[dst]) return result;

    result.reachable = true;
    int cur = dst;
    while (cur != -1) {
        result.path.push_back(cur);
        cur = prev[cur];
    }
    std::reverse(result.path.begin(), result.path.end());

    // Sum actual costs along the BFS path
    for (size_t i = 0; i + 1 < result.path.size(); i++) {
        int from = result.path[i];
        int to = result.path[i + 1];
        for (const auto& e : adj_[from]) {
            if (e.to == to) {
                result.total_cost += e.cost;
                result.total_time += e.time_hours;
                break;
            }
        }
    }

    return result;
}

// ──────────────────────────────────────────────
// All nodes within cost threshold using modified Dijkstra
// ──────────────────────────────────────────────
std::vector<std::pair<int, double>> Graph::withinCostThreshold(int src, double maxCost) const {
    std::vector<std::pair<int, double>> result;
    int n = (int)adj_.size();
    if (src < 0 || src >= n) return result;

    const double INF = std::numeric_limits<double>::infinity();
    std::vector<double> dist(n, INF);

    using PII = std::pair<double, int>;
    std::priority_queue<PII, std::vector<PII>, std::greater<PII>> pq;

    dist[src] = 0;
    pq.push({0, src});

    while (!pq.empty()) {
        auto topPair = pq.top();
        double d = topPair.first;
        int u = topPair.second;
        pq.pop();
        if (d > dist[u]) continue;
        if (d > maxCost) break;

        if (u != src && warehouses_.count(u)) {
            result.push_back({u, d});
        }

        for (const auto& edge : adj_[u]) {
            double nd = dist[u] + edge.cost;
            if (nd < dist[edge.to] && nd <= maxCost) {
                dist[edge.to] = nd;
                pq.push({nd, edge.to});
            }
        }
    }

    return result;
}

// ──────────────────────────────────────────────
// CSV Loaders
// ──────────────────────────────────────────────
void Graph::loadWarehouses(const std::string& path) {
    std::ifstream file(path);
    std::string line;
    std::getline(file, line); // skip header

    while (std::getline(file, line)) {
        std::istringstream ss(line);
        std::string token;
        Warehouse w;

        std::getline(ss, token, ','); w.id = std::stoi(trim(token));
        std::getline(ss, token, ','); w.name = trim(token);
        std::getline(ss, token, ','); w.city = trim(token);
        std::getline(ss, token, ','); w.lat = std::stod(trim(token));
        std::getline(ss, token, ','); w.lng = std::stod(trim(token));
        std::getline(ss, token, ','); w.capacity = std::stoi(trim(token));

        addWarehouse(w);
    }
}

void Graph::loadRoutes(const std::string& path) {
    std::ifstream file(path);
    std::string line;
    std::getline(file, line); // skip header

    while (std::getline(file, line)) {
        std::istringstream ss(line);
        std::string token;

        int from, to;
        double cost, time_hours;

        std::getline(ss, token, ','); from = std::stoi(trim(token));
        std::getline(ss, token, ','); to = std::stoi(trim(token));
        std::getline(ss, token, ','); cost = std::stod(trim(token));
        std::getline(ss, token, ','); time_hours = std::stod(trim(token));
        // skip distance_km col

        addEdge(from, to, cost, time_hours);
    }
}

File: engine/src/main/java/com/optiflow/engine/Graph.java
package com.optiflow.engine;

import java.io.*;
import java.util.*;

/**
 * ──────────────────────────────────────────────
 * Logistics Graph Engine
 *   - Adjacency List representation
 *   - Dijkstra's shortest path: O((V+E) log V)
 *   - BFS shortest hop path:    O(V+E)
 * ──────────────────────────────────────────────
 */
public class Graph {

    // ── Data Classes ──

    public static class Edge {
        public int to;
        public double cost;       // delivery cost in ₹
        public double timeHours;  // transit time

        public Edge(int to, double cost, double timeHours) {
            this.to = to;
            this.cost = cost;
            this.timeHours = timeHours;
        }
    }

    public static class PathResult {
        public double totalCost;
        public double totalTime;
        public List<Integer> path;  // sequence of warehouse IDs
        public boolean reachable;

        public PathResult() {
            this.totalCost = 0;
            this.totalTime = 0;
            this.path = new ArrayList<>();
            this.reachable = false;
        }
    }

    public static class Warehouse {
        public int id;
        public String name;
        public String city;
        public double lat, lng;
        public int capacity;
    }

    // ── Fields ──

    private final Map<Integer, Warehouse> warehouses = new LinkedHashMap<>();
    private final List<List<Edge>> adj = new ArrayList<>();
    private int maxId = 0;

    // ── Graph Construction ──

    private void ensureSize(int id) {
        if (id > maxId) {
            maxId = id;
            while (adj.size() <= maxId) {
                adj.add(new ArrayList<>());
            }
        }
    }

    public void addWarehouse(Warehouse w) {
        ensureSize(w.id);
        warehouses.put(w.id, w);
    }

    public void addEdge(int from, int to, double cost, double timeHours) {
        ensureSize(from);
        ensureSize(to);
        // Bidirectional edges
        adj.get(from).add(new Edge(to, cost, timeHours));
        adj.get(to).add(new Edge(from, cost, timeHours));
    }

    // ── Dijkstra's Algorithm — O((V+E) log V) ──
    //   Uses min-heap (priority queue) to greedily
    //   expand the cheapest frontier node.

    public PathResult dijkstra(int src, int dst) {
        PathResult result = new PathResult();

        int n = adj.size();
        if (src < 0 || src >= n || dst < 0 || dst >= n) return result;

        double[] dist = new double[n];
        double[] timeArr = new double[n];
        int[] prev = new int[n];
        Arrays.fill(dist, Double.MAX_VALUE);
        Arrays.fill(timeArr, Double.MAX_VALUE);
        Arrays.fill(prev, -1);

        // Min-heap: (cost, node)
        PriorityQueue<double[]> pq = new PriorityQueue<>(Comparator.comparingDouble(a -> a[0]));

        dist[src] = 0;
        timeArr[src] = 0;
        pq.offer(new double[]{0, src});

        while (!pq.isEmpty()) {
            double[] top = pq.poll();
            double d = top[0];
            int u = (int) top[1];

            if (d > dist[u]) continue;  // stale entry
            if (u == dst) break;        // found shortest

            for (Edge edge : adj.get(u)) {
                double newDist = dist[u] + edge.cost;
                if (newDist < dist[edge.to]) {
                    dist[edge.to] = newDist;
                    timeArr[edge.to] = timeArr[u] + edge.timeHours;
                    prev[edge.to] = u;
                    pq.offer(new double[]{newDist, edge.to});
                }
            }
        }

        if (dist[dst] == Double.MAX_VALUE) return result;

        // Reconstruct path by backtracking through prev[]
        result.reachable = true;
        result.totalCost = dist[dst];
        result.totalTime = timeArr[dst];

        int cur = dst;
        while (cur != -1) {
            result.path.add(cur);
            cur = prev[cur];
        }
        Collections.reverse(result.path);

        return result;
    }

    // ── BFS — O(V+E) shortest path by hop count ──
    //   Treats all edges as weight=1

    public PathResult bfs(int src, int dst) {
        PathResult result = new PathResult();

        int n = adj.size();
        if (src < 0 || src >= n || dst < 0 || dst >= n) return result;

        int[] prev = new int[n];
        boolean[] visited = new boolean[n];
        Arrays.fill(prev, -1);

        visited[src] = true;
        Queue<Integer> q = new LinkedList<>();
        q.add(src);

        while (!q.isEmpty()) {
            int u = q.poll();
            if (u == dst) break;

            for (Edge edge : adj.get(u)) {
                if (!visited[edge.to]) {
                    visited[edge.to] = true;
                    prev[edge.to] = u;
                    q.add(edge.to);
                }
            }
        }

        if (!visited[dst]) return result;

        result.reachable = true;
        int cur = dst;
        while (cur != -1) {
            result.path.add(cur);
            cur = prev[cur];
        }
        Collections.reverse(result.path);

        // Sum actual costs along the BFS path
        for (int i = 0; i + 1 < result.path.size(); i++) {
            int from = result.path.get(i);
            int to = result.path.get(i + 1);
            for (Edge e : adj.get(from)) {
                if (e.to == to) {
                    result.totalCost += e.cost;
                    result.totalTime += e.timeHours;
                    break;
                }
            }
        }

        return result;
    }

    // ── All nodes within cost threshold using modified Dijkstra ──

    public List<int[]> withinCostThreshold(int src, double maxCost) {
        List<int[]> result = new ArrayList<>();
        int n = adj.size();
        if (src < 0 || src >= n) return result;

        double[] dist = new double[n];
        Arrays.fill(dist, Double.MAX_VALUE);

        PriorityQueue<double[]> pq = new PriorityQueue<>(Comparator.comparingDouble(a -> a[0]));
        dist[src] = 0;
        pq.offer(new double[]{0, src});

        while (!pq.isEmpty()) {
            double[] top = pq.poll();
            double d = top[0];
            int u = (int) top[1];
            if (d > dist[u]) continue;
            if (d > maxCost) break;

            if (u != src && warehouses.containsKey(u)) {
                result.add(new int[]{u, (int) d});
            }

            for (Edge edge : adj.get(u)) {
                double nd = dist[u] + edge.cost;
                if (nd < dist[edge.to] && nd <= maxCost) {
                    dist[edge.to] = nd;
                    pq.offer(new double[]{nd, edge.to});
                }
            }
        }

        return result;
    }

    // ── CSV Loaders ──

    private static String trim(String s) {
        return s == null ? "" : s.trim();
    }

    public void loadWarehouses(String path) throws IOException {
        try (BufferedReader br = new BufferedReader(new FileReader(path))) {
            br.readLine(); // skip header
            String line;
            while ((line = br.readLine()) != null) {
                String[] tokens = line.split(",");
                Warehouse w = new Warehouse();
                w.id = Integer.parseInt(trim(tokens[0]));
                w.name = trim(tokens[1]);
                w.city = trim(tokens[2]);
                w.lat = Double.parseDouble(trim(tokens[3]));
                w.lng = Double.parseDouble(trim(tokens[4]));
                w.capacity = Integer.parseInt(trim(tokens[5]));
                addWarehouse(w);
            }
        }
    }

    public void loadRoutes(String path) throws IOException {
        try (BufferedReader br = new BufferedReader(new FileReader(path))) {
            br.readLine(); // skip header
            String line;
            while ((line = br.readLine()) != null) {
                String[] tokens = line.split(",");
                int from = Integer.parseInt(trim(tokens[0]));
                int to = Integer.parseInt(trim(tokens[1]));
                double cost = Double.parseDouble(trim(tokens[2]));
                double timeHours = Double.parseDouble(trim(tokens[3]));
                addEdge(from, to, cost, timeHours);
            }
        }
    }

    // ── Accessors ──

    public Map<Integer, Warehouse> getWarehouses() {
        return warehouses;
    }

    public List<List<Edge>> getAdj() {
        return adj;
    }

    public int nodeCount() {
        return maxId + 1;
    }
}
